proto/api/v1/notification_service.proto

```diff
@@ -99,19 +75,71 @@ enum NotificationLevel {
     NOTIFICATION_LEVEL_DANGER  = 4;
 }
 
+message NotificationsFilter {
+    // For filtering notifications starting from a specific timestamp, i.e. the lower boundary.
+    google.protobuf.Timestamp from     = 1;
+
+    // For filtering notifications up until a specific timestamp, i.e. the upper boundary.
+    google.protobuf.Timestamp until    = 2;
+
+    // For filtering notifications from a specific domain.
+    string domain                      = 3;
+
+    // For filtering notifications associated with a specific resource type.
+    string resource_type               = 4;
+
+    // For filtering notifications based on their type.
+    NotificationType notification_type = 5;
+
+    // For filtering notifications based on their level.
+    NotificationLevel level            = 6;
+}
+
+message CountNotificationsRequest {
+    // For filtering the notifications based on the requested fields.
+    NotificationsFilter filter = 1;
+}
+
+message CountNotificationsResponse {
+    // The total number of notifications after filtering and deduplication.
+    int64 count = 1;
+}
+
+message GetNotificationResponse {
+    Notification notification = 1;
+}
+
+message ListNotificationsRequest {
+    // For dividing the notifications response into chunks.
+    Pagination pagination      = 1;
+
+    // For filtering the notifications based on the requested fields.
+    NotificationsFilter filter = 2;
+}
+
+message ListNotificationsResponse {
+    repeated Notification notifications = 1;
+}
+
 service NotificationService {
-    // GetNotifications returns the list of notifications. The events may be filtered by
-    // dedicated fields.
-    rpc GetNotifications(GetNotificationsRequest) returns (NotificationsResponse) {
+    // CountNotifications returns the number of notifications after filtered by requested fields.
+    rpc CountNotifications(CountNotificationsRequest) returns (CountNotificationsResponse) {
         option (google.api.http) = {
-            get: "/v1/notifications"
+            get: "/v1/notificationscount"
         };
     }
 
     // GetNotification retrieves a notification by ID.
-    rpc GetNotification(ResourceByID) returns (Notification) {
+    rpc GetNotification(ResourceByID) returns (GetNotificationResponse) {
         option(google.api.http) = {
             get: "/v1/notifications/{id}"
         };
     }
+
+    // ListNotifications returns the list of notifications after filtered by requested fields.
+    rpc ListNotifications(ListNotificationsRequest) returns (ListNotificationsResponse) {
```

Comment:
Are these APIs externally exposed?  Is the data returned exactly the same way it is today?  My concern is that it may be nested in another level of object so these changes may break the compatibility of the API.

Reply:
Not yet, see later PRs that define the service layer.